wsd/Storage.cpp

```diff
@@ -330,25 +317,55 @@ std::string LocalStorage::loadStorageFileToLocal(const Authorization& /*auth*/,
 
     // Despite the talk about URIs it seems that _uri is actually just a pathname here
     const std::string publicFilePath = getUri().getPath();
+    if (!Poco::File(publicFilePath).exists())
+    {
+        LOG_ERR("Local file URI [" << publicFilePath << "] invalid or doesn't exist.");
+        throw BadRequestException("Invalid URI: " + getUri().toString());
+    }
 
     if (!FileUtil::checkDiskSpace(getRootFilePath()))
     {
         throw StorageSpaceLowException("Low disk space for " + getRootFilePathAnonym());
     }
 
-    LOG_INF("Linking " << LOOLWSD::anonymizeUrl(publicFilePath) << " to " << getRootFilePathAnonym());
-    if (!Poco::File(getRootFilePath()).exists() && link(publicFilePath.c_str(), getRootFilePath().c_str()) == -1)
+    if (_isTemporaryFile)
+    {
+        try
+        {
+            // Neither link nor copy, just move, it's a temporary file.
+            Poco::File(publicFilePath).moveTo(getRootFilePath());
```

Comment:
This is the key part: by moving the doc into the destination path where it's loaded from, there is no temp file to cleanup: the jail is cleaned up when the document is unloaded.